File: src/test/java/com/matamercer/microblog/integration/controller/LoginControllerTest.java
package com.matamercer.microblog.integration.controller;

import com.matamercer.microblog.forms.LoginUserForm;
import com.matamercer.microblog.models.entities.AuthenticationProvider;
import com.matamercer.microblog.models.entities.User;
import com.matamercer.microblog.models.repositories.UserRepository;
import com.matamercer.microblog.security.UserRole;
import com.matamercer.microblog.security.jwt.JwtConfig;
import com.matamercer.microblog.security.jwt.JwtUtil;
import com.matamercer.microblog.services.UserService;
import com.matamercer.microblog.utilities.EnvironmentUtil;
import lombok.val;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.RequestEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.test.context.junit4.SpringRunner;

import java.net.URI;
import java.net.UnknownHostException;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class LoginControllerTest {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private UserService userService;

    @Autowired
    private EnvironmentUtil environmentUtil;

    @Autowired
    private JwtUtil jwtUtil;

    @Autowired
    private JwtConfig jwtConfig;

    private TestRestTemplate testRestTemplate;

    private User user;

    private String unencodedPassword;



    @Before
    public void initData(){
        unencodedPassword = "password";
        user = new User("username@gmail.com",
                "username",
                passwordEncoder.encode(unencodedPassword),
                UserRole.USER,
                true,
                true,
                true,
                true,
                AuthenticationProvider.LOCAL);
        user = userService.createUser(user);
        testRestTemplate = new TestRestTemplate();
    }

    @After
    public void disposeData(){
        userRepository.deleteById(user.getId());
    }


    private RequestEntity<LoginUserForm> getLoginRequestEntity(LoginUserForm loginUserForm) throws UnknownHostException {
        return RequestEntity
                .post(URI.create(environmentUtil.getServerUrl()+ "/login"))
                .contentType(MediaType.APPLICATION_JSON)
                .body(loginUserForm);
    }

    private RequestEntity<Void> getCurrentUserRequestEntity(String accessToken) throws UnknownHostException {
        HttpHeaders headers = new HttpHeaders();
        headers.add(jwtConfig.getAuthorizationHeader(), accessToken);
        return RequestEntity.get(URI.create(environmentUtil.getServerUrl() + "/currentuser"))
                .headers(headers)
                .build();
    }


    @Test
    public void whenLogin_withGoodCredentials_thenReturnResponseIsOk_WithAccessAndRefreshTokens() throws UnknownHostException {
        val requestEntity = getLoginRequestEntity(new LoginUserForm(user.getUsername(), unencodedPassword));
        val responseEntity = testRestTemplate.exchange(requestEntity, Object.class);

        assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
        val accessToken = responseEntity.getHeaders().get(jwtConfig.getAuthorizationHeader()).stream().findFirst();
        assertThat(accessToken.isPresent()).isTrue();
        val refreshToken = responseEntity.getHeaders().get(jwtConfig.getRefreshTokenHeader()).stream().findFirst();
        assertThat(refreshToken.isPresent()).isTrue();
    }

    @Test
    public void whenGetCurrentUser_withAccessToken_thenReturnCurrentUser() throws UnknownHostException {
        val requestEntity = getLoginRequestEntity(new LoginUserForm(user.getUsername(), unencodedPassword));
        val responseEntity = testRestTemplate.exchange(requestEntity, Object.class);

        assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
        val accessToken = responseEntity.getHeaders().get(jwtConfig.getAuthorizationHeader()).stream().findFirst();
        assertThat(accessToken.isPresent()).isTrue();
        val refreshToken = responseEntity.getHeaders().get(jwtConfig.getRefreshTokenHeader()).stream().findFirst();
        assertThat(refreshToken.isPresent()).isTrue();

        val currentUserRequestEntity = getCurrentUserRequestEntity(accessToken.get());
        val currentUserResponseEntity= testRestTemplate.exchange(currentUserRequestEntity, String.class);

        assertThat(currentUserResponseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
    }

    @Test
    public void whenGetCurrentUser_withoutAccessToken_thenReturnCurrentUser() throws UnknownHostException {

        val currentUserRequestEntity = getCurrentUserRequestEntity("");
        val currentUserResponseEntity= testRestTemplate.exchange(currentUserRequestEntity, String.class);

        assertThat(currentUserResponseEntity.getStatusCode()).isEqualTo(HttpStatus.FORBIDDEN);
    }

}

File: src/test/java/com/matamercer/microblog/integration/controller/LoginControllerTest.kt
package com.matamercer.microblog.integration.controller

import com.matamercer.microblog.models.entities.AuthenticationProvider
import com.matamercer.microblog.models.entities.Blog
import com.matamercer.microblog.models.entities.User
import com.matamercer.microblog.models.repositories.BlogRepository
import com.matamercer.microblog.models.repositories.RefreshTokenRepository
import com.matamercer.microblog.models.repositories.UserRepository
import com.matamercer.microblog.security.authentication.JwtConfig
import com.matamercer.microblog.security.authorization.UserRole
import com.matamercer.microblog.services.UserService
import com.matamercer.microblog.utilities.EnvironmentUtil
import com.matamercer.microblog.web.api.v1.dto.requests.LoginUserRequestDto
import io.mockk.clearAllMocks
import io.mockk.junit5.MockKExtension
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.*
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.net.URI
import java.net.UnknownHostException

@ExtendWith(MockKExtension::class, SpringExtension::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ActiveProfiles("test")
class LoginControllerTest @Autowired constructor(
    private val userService: UserService,
    private val passwordEncoder: PasswordEncoder,
    private val jwtConfig: JwtConfig,
    private val environmentUtil: EnvironmentUtil,
    private val userRepository: UserRepository,
    private val refreshTokenRepository: RefreshTokenRepository,
    private val blogRepository: BlogRepository
) {
    private var testRestTemplate: TestRestTemplate? = null

    private lateinit var user: User

    private lateinit var responseEntity: ResponseEntity<Any>

    private lateinit var requestEntity: RequestEntity<LoginUserRequestDto>

    private var unencodedPassword: String? = null

    @BeforeEach
    fun initData() {
        clearAllMocks()
        userRepository.deleteAll()
        refreshTokenRepository.deleteAll()
        blogRepository.deleteAll()
        unencodedPassword = "password"
        user = User(
            email = "username@gmail.com",
            username = "username",
            password = passwordEncoder.encode(unencodedPassword),
            role = UserRole.BLOGGER,
            authenticationProvider = AuthenticationProvider.LOCAL
        )
        val blog = Blog(isSensitive = true, preferredBlogName = "Testermourne", blogName = "Tester Agony")
        blogRepository.save(blog)
        user.activeBlog = blog
        user = userService.createUser(user)
        testRestTemplate = TestRestTemplate()
        requestEntity = getLoginRequestEntity(
            LoginUserRequestDto(
                user.username!!,
                unencodedPassword!!
            )
        )
        responseEntity = testRestTemplate!!.exchange(requestEntity, Any::class.java)
    }

    @Throws(UnknownHostException::class)
    private fun getLoginRequestEntity(loginUserRequestDTO: LoginUserRequestDto): RequestEntity<LoginUserRequestDto> {
        return RequestEntity
            .post(URI.create(environmentUtil.serverUrl + "/api/v1/auth/login"))
            .contentType(MediaType.APPLICATION_JSON)
            .body(loginUserRequestDTO)
    }

    @Throws(UnknownHostException::class)
    private fun getCurrentUserRequestEntity(accessToken: String): RequestEntity<Void> {
        val headers = HttpHeaders()
        headers.add(jwtConfig.authorizationHeader, accessToken)
        return RequestEntity.get(URI.create(environmentUtil.serverUrl + "/api/v1/user/currentuser"))
            .headers(headers)
            .build()
    }

    @Test
    @Throws(UnknownHostException::class)
    fun whenLogin_withGoodCredentials_thenReturnResponseIsOk_WithAccessAndRefreshTokens() {
        assertThat(responseEntity.statusCode).isEqualTo(HttpStatus.OK)
        val accessToken = responseEntity.headers[jwtConfig.authorizationHeader]!!.first()
        assertThat(accessToken != null).isTrue
        val refreshToken = responseEntity.headers[jwtConfig.refreshTokenHeader]!!.first()
        assertThat(refreshToken != null).isTrue
    }

    @Test
    @Throws(UnknownHostException::class)
    fun whenGetCurrentUser_withAccessToken_thenReturnCurrentUser() {
        val accessToken = responseEntity.headers[jwtConfig.authorizationHeader]!!.first()
        assertThat(accessToken != null).isTrue
        val refreshToken = responseEntity.headers[jwtConfig.refreshTokenHeader]!!.first()
        assertThat(refreshToken != null).isTrue
        val currentUserRequestEntity = getCurrentUserRequestEntity(accessToken)
        val currentUserWithTokenResponseEntity = testRestTemplate!!.exchange(
            currentUserRequestEntity,
            String::class.java
        )
        assertThat(currentUserWithTokenResponseEntity.statusCode).isEqualTo(HttpStatus.OK)
    }

    @Test
    @Throws(UnknownHostException::class)
    fun whenGetCurrentUser_withoutAccessToken_thenReturnCurrentUser() {
        val currentUserWithOutTokenResponseEntity = testRestTemplate!!.exchange(
            getCurrentUserRequestEntity(""),
            String::class.java
        )
        assertThat(currentUserWithOutTokenResponseEntity.statusCode).isEqualTo(HttpStatus.FORBIDDEN)
    }
}